Implement RFC#1559: allow all literals in attributes

SergioBenitez opened the pull request:
Implemented rust-lang/rfcs#1559, tracked by #34981.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @nikomatsakis (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

eddyb commented on src/librustc/hir/check_attr.rs:
```diff
@@ -53,33 +53,36 @@ impl<'a> CheckAttrVisitor<'a> {
-                        continue
+            let name = word.name();
+            let message = match name {
+                Some(word) => match &*word {
```
This could probably be `if let Some(word) = word.name() {`.

eddyb replied:
Wait, doesn't `.name()` cause an error for `repr(5)` currently, that would be silenced?

SergioBenitez replied:
I'll use `if let ...`. The check you're referring to happens [https://github.com/rust-lang/rust/pull/35850/files#diff-21e1f9f42b7976b50c28979848df2e04R951](here).

SergioBenitez replied:
Actually, it's slightly cleaner to use `match word.name()`, so I'll switch it to that.

eddyb replied:
Might be able to do `match word.name().as_ref().map_or("", |s| s)` (or `map_or("", |s| &**s)`) - assuming that it has been previously validated.

SergioBenitez replied:
Yeah, I considered that, but we don't know if it's a word at this point (though it doesn't matter too much, yet). Plus, it seems a bit more obvious to specifically match on the None case.

eddyb replied:
Two nested `match` are pretty ugly :disappointed:.

SergioBenitez replied:
Changed to (from IRC):

```
            let name = match word.name() {
                Some(word) => word,
                None => continue,
            };

            let message = match &*name {
                ...
            }

            span_err!(self.sess, attr.span, E0517, "{}", message);
```

eddyb commented on src/librustc_metadata/common.rs:
```diff
@@ -50,21 +50,8 @@ pub const tag_def_key: usize = 0x2c;
-pub const tag_meta_item_word: usize = 0x33;
-
-pub const tag_meta_item_list: usize = 0x34;
-
```
These should use the `GAP` comments.

SergioBenitez replied:
Got it. Will add.

eddyb commented on src/libsyntax/ast.rs:
```diff
@@ -1105,6 +1123,26 @@ impl LitKind {
+            LitKind::Int(_, LitIntType::Unsuffixed) => true,
+            LitKind::FloatUnsuffixed(..) => true,
+            LitKind::Bool(..) => true,
+            _ => false,
```
Maybe all the cases should be written out, to avoid problems in the future.

SergioBenitez replied:
That's a good idea. I'll do that.

cgswords replied:
Why add this here, and not to the LitKind impl?

SergioBenitez replied:
This is the LitKind impl...

cgswords replied:
Whoops, nevermind. I was thinking of the `Lit` stuff in `token`; disregard!

eddyb commented on src/libsyntax/parse/attr.rs:
```diff
@@ -193,9 +193,26 @@ impl<'a> Parser<'a> {
+        debug!("Checking if {:?} is unusuffixed.", lit);
+
+        if !lit.node.is_unsuffixed() {
+            let msg = "suffixed literals are not allowed in meta-items";
```
I don't think we should leak "meta-items" to the user. "in attributes" still seems correct to me and pretty clear.

SergioBenitez replied:
Sounds good to me.

eddyb commented:
Looks pretty good to me, great job! :tada:

cc @nrc @cgswords @Manishearth

Mark-Simulacrum commented on src/librustc_incremental/assert_dep_graph.rs:
```diff
@@ -116,31 +116,38 @@ impl<'a, 'tcx> IfThisChanged<'a, 'tcx> {
-                    if meta_item.is_word() && id.is_none() {
-                        id = Some(meta_item.name().clone());
-                    } else {
-                        // FIXME better-encapsulate meta_item (don't directly access `node`)
```
Node is still directly accessed in the span_bug! so perhaps the FIXME shouldn't be removed?

SergioBenitez replied:
This happens in a lot of places without a similar comment. Is it more important to place such a comment here than in other places?

cgswords replied:
I added those fixmes; ultimately, there should be some encapsulation to make meta-items more opaque (so that their internal representation is less obvious and more resilient to changes). If it's done other places, this fixme should be added there, too.

Mark-Simulacrum commented on src/librustc_incremental/assert_dep_graph.rs:
```diff
@@ -116,31 +116,38 @@ impl<'a, 'tcx> IfThisChanged<'a, 'tcx> {
-                    } else if meta_item.is_word() && id.is_none() {
-                        id = Some(meta_item.name().clone());
-                    } else {
-                        // FIXME better-encapsulate meta_item (don't directly access `node`)
```
Same as above; FIXME might still need to be kept.

SergioBenitez replied:
Okay, I'll re-add them.

cgswords commented on src/libsyntax/parse/attr.rs:
```diff
@@ -213,36 +230,43 @@ impl<'a> Parser<'a> {
-                                      "non-string literals are not allowed in meta-items");
-                    }
-                }
+                let lit = self.parse_unsuffixed_lit()?;
```
Why has this been changed to be an unsuffixed literal? Supporting all literals means supporting suffixed literals (and the parser already has that operation).

SergioBenitez replied:
That was the conclusion we came to in https://github.com/rust-lang/rfcs/pull/1559,

cgswords replied:
Well okay, then. :cry:

SergioBenitez replied:
Would be a one line change (+ some deletions) to reconsider in the future, however!

nrc replied:
FWIW, I think we should accept suffixed literals, I don't understand the argument against.

cgswords commented on src/libsyntax/parse/attr.rs:
```diff
@@ -213,36 +230,43 @@ impl<'a> Parser<'a> {
                 let hi = self.span.hi;
                 Ok(P(spanned(lo, hi, ast::MetaItemKind::NameValue(name, lit))))
-            }
+            },
```
Nit: we don't need this comma, get rid of it (and same with the one added below)

SergioBenitez replied:
While we don't need it, I believe this style to be less error prone (if the arm gets changed). It also appears to be the style used in much of the code.

cgswords replied:
`rustfmt` will remove the commas, and it is generally used to inform the compiler style guide

SergioBenitez replied:
Sounds good. Will do.

cgswords commented on src/libsyntax/ast.rs:
```diff
@@ -456,7 +472,7 @@ pub enum MetaItemKind {
     ///
     /// E.g. `derive(..)` as in `#[derive(..)]`
-    List(InternedString, Vec<P<MetaItem>>),
+    List(InternedString, Vec<MetaListItem>),
```
I am really dubious about removing this pointer. It means we're going to clone a goodly number of literals, which might cause perf regressions in, eg, serde.

eddyb replied:
`P` is owned though, so they wouldn't be shared, unless you mean just moving the value around? Not only that, but ASTs shouldn't get cloned all the time.

cgswords replied:
Yeah, sorry, that is what I meant (though iirc meta-items do end up getting cloned here and there).

SergioBenitez replied:
Can we land this and benchmark?

cgswords replied:
It's preferable to benchmark before landing, because otherwise it will mean immediately landing a second PR just to fix it. `rustup` will let you use a local build as the default; you can build `serde` and run its perf benchmarks with both.

SergioBenitez replied:
Thinking about this: MetaItem almost always sites behind a P<> anyways, so there should be very little, if any, performance degradation. We might even see a performance win once literals start being use due to coalesced allocations.

cgswords commented on src/libsyntax/attr.rs:
```diff
@@ -77,6 +81,93 @@ pub fn is_used(attr: &Attribute) -> bool {
+pub trait AttrMetaListItemMethods {
+    /// Returns true if this list item is a MetaItem with a name of `name`.
+    fn check_name(&self, name: &str) -> bool {
+        self.meta_item().map_or(false, |meta_item| meta_item.check_name(name))
```
Nit: You probably want `and_then`.

cgswords commented on src/libsyntax/attr.rs:
```diff
@@ -77,6 +81,93 @@ pub fn is_used(attr: &Attribute) -> bool {
     })
 }
 
+pub trait AttrMetaListItemMethods {
```
It seems like adding `literal` to `AttrMetaMethods` will be a lot cleaner than defining an entire duplication of `AttrMetaMethods` to add `literal`. It will also make things more uniform in the other parts of the compiler.

SergioBenitez replied:
The issues is that AttrMetaMethod's `name()` always returns an InternedString, while ListMetaMethods returns an Option<...>.

cgswords replied:
`ListMetaMethods` should probably return the literal as a string for `name()` instead of a `None` (cc @nrc)

SergioBenitez replied:
My opinion is that since this is not the way `name()` is used in the code, it would be error prone to make this change.

nrc replied:
Doesn't sounds like there is a good solution here, but at the least something needs renaming or there needs to be some docs here - totally unclear to me what the abstractions here represent

SergioBenitez replied:
Discussed with @nrc over IRC. Conclusion was that we'll keep the two traits since MetaItem and MetaListItems are logically different.

cgswords commented on src/libsyntax/attr.rs:
```diff
@@ -89,8 +180,9 @@ pub trait AttrMetaMethods {
+
     /// Gets a list of inner meta items from a list MetaItem type.
-    fn meta_item_list(&self) -> Option<&[P<MetaItem>]>;
+    fn meta_item_list(&self) -> Option<&[MetaListItem]>;
```
Removing the pointer here (+below) may cause perf regressions with serde, etc.

SergioBenitez replied:
Same as above.

cgswords commented on src/libsyntax/attr.rs:
```diff
@@ -557,14 +689,19 @@ fn find_stability_generic<'a, I>(diagnostic: &Handler,
-                                handle_errors(diagnostic, meta.span,
-                                              AttrError::UnknownMetaItem(meta.name()));
-                                continue 'outer
+                        if let MetaListItemKind::MetaItem(ref mi) = meta.node {
```
Please write a variant of `list_contains_name` (or change its type signature) to return an `Option`, then rewrite this (and maybe also `cfg_matches` to use that instead of the mess that's here right now.

SergioBenitez replied:
Oh, that already exists. This just doesn't use it for some reason.

cgswords commented on src/libsyntax/test.rs:
```diff
@@ -19,8 +19,7 @@ use std::iter;
 use std::vec;
-use attr::AttrMetaMethods;
-use attr;
+use attr::{self, AttrMetaMethods, AttrMetaListItemMethods};
```
Please add some tests explicitly for MetaItemLists here.

cgswords commented on src/libsyntax/attr.rs:
```diff
@@ -453,26 +573,38 @@ pub fn requests_inline(attrs: &[Attribute]) -> bool {
 /// Tests if a cfg-pattern matches the cfg set
 pub fn cfg_matches(cfgs: &[P<MetaItem>], cfg: &ast::MetaItem,
-                   sess: &ParseSess, features: Option<&Features>)
+                   sess: &ParseSess, fs: Option<&Features>)
```
Nit: don't make the variable names worse.

SergioBenitez replied:
This was so line 590/91 would be <= 100 cols.

cgswords replied:
I'm pretty sure this doesn't adhere to what `rustfmt` would do, anyway. It should be:

```rust
pub fn cfg_matches(cfgs: &[P<MetaItem>],
                   cfg: &ast::MetaItem,
                   sess: &ParseSess,
                   features: Option<&Features>)
                   -> bool {
```

SergioBenitez replied:
I'll split the "any" and "all" branches into multiple lines and keep `features`.

cgswords commented on src/libsyntax/attr.rs:
```diff
@@ -229,10 +324,14 @@ pub fn mk_name_value_item(name: InternedString, value: ast::Lit)
 }
 
+pub fn mk_list_word_item(name: InternedString) -> ast::MetaListItem {
+    respan(DUMMY_SP, MetaListItemKind::MetaItem(mk_spanned_word_item(DUMMY_SP, name)))
```
Nit: use `dummy_spanned` instead of `respan` here

SergioBenitez replied:
Sure thing.

cgswords commented on src/librustc_driver/lib.rs:
```diff
@@ -666,6 +666,9 @@ impl RustcDefaultCalls {
                             panic!("MetaItemKind::List encountered in default cfg")
+                        } else {
+                            // There also shouldn't be literals.
+                            panic!("Unexpected MetaItem literal in default cfg!")
```
Maybe panic about a malformed attribute, not a meta-item.

SergioBenitez replied:
The panic above mentions meta item, so I thought this would be more uniform.

cgswords replied:
Change both? It would be nice to give more descriptive errors with some of the values on-hand, too (eg emitting the `config` name that caused the problem).

SergioBenitez replied:
Sounds good.

cgswords commented:
There should be some `run-pass` tests.

SergioBenitez commented on src/libsyntax/ast.rs:
```diff
@@ -1105,6 +1123,26 @@ impl LitKind {
+    }
+
+    /// Returns true if this literal has a suffix.
+    pub fn is_suffixed(&self) -> bool {
```
See above.

cgswords commented:
Aside from my comments and request for tests, this is basically all awesome. :tada:

Manishearth commented:
When this is ready to merge, let me know, we can batch up everything in https://github.com/rust-lang/rust/issues/31645#issuecomment-239228452

nrc commented on src/librustc_borrowck/borrowck/mir/mod.rs:
```diff
@@ -43,8 +43,10 @@ fn has_rustc_mir_with(attrs: &[ast::Attribute], name: &str) -> Option<P<MetaItem
+                if let Some(mi) = item.meta_item() {
+                    if mi.check_name(name) {
+                        return Some(mi.clone())
+                    }
```
Needing double `if`s here is annoying - could check_name return None for non-meta items? (More generally, I kind of hate the concept of meta-item - it is arbitrary and meaningless. We should do our best not to expose the concept outside of libsyntax).

SergioBenitez replied:
Actually, we can call check_name directly on the list item. I'll change it to that.

nrc commented on src/librustc_driver/lib.rs:
```diff
@@ -655,17 +655,19 @@ impl RustcDefaultCalls {
+                            panic!("Found an unexpected list in cfg attribute '{}'!", cfg.name())
+                        } else {
+                            // There also shouldn't be literals.
+                            panic!("Found an unexpected literal in cfg attribute '{}'!", cfg.name())
```
Given that this is not meant to be a user-facing error message, I don't think we need two different messages

SergioBenitez replied:
The two different error messages should help out if the conditions are ever hit. I propose we keep them: doing so doesn't hurt and can only help.

nrc commented on src/libsyntax/ast.rs:
```diff
@@ -439,6 +439,22 @@ pub struct Crate {
 }
 
+/// A spanned compile-time attribute list item.
+pub type MetaListItem = Spanned<MetaListItemKind>;
```
Maybe `NestedMetaItem` - the list part of the name is a bit misleading

SergioBenitez replied:
A `MetaItem` can already be nested in itself, but it cannot appear in a list. Thus, the ability to nest isn't the differentiation, it's the fact that the type can appear inside of a list. If others feels strongly about this, I'm okay with changing the name.

nrc replied:
I would like the name changes, MetaListItem doesn't seem very informative to me

nrc commented:
This should be feature-gated, right?

SergioBenitez commented:
Why should this be feature gated? The syntax is backwards compatible, and custom attributes are nightly only.

eddyb commented:
@SergioBenitez So you're saying all stable attributes do checks that prevent use of literals?

SergioBenitez commented:
@eddyb: That's the intention with this PR, yeah.

nrc commented:
I would be more paranoid than that - there is nothing to stop someone accidentally accepting a literal in a stable attribute or adding a new attribute which requires a literal.

eddyb commented on src/test/run-pass-fulldeps/auxiliary/macro_crate_test.rs:
```diff
@@ -42,40 +44,45 @@ pub fn plugin_registrar(reg: &mut Registry) {
     match it {
-        Annotatable::Item(it) => vec![
+        Annotatable::Item(it) => {
+            vec![
```
Why this change? For the indentation to match up you'd need to indent all the lines inside - it's simpler just to keep it the way it is.

SergioBenitez replied:
This was `rustfmt`. Happy to change it back.

eddyb replied:
cc @nikomatsakis @ubsan Another case where rustfmt is ugly *and* wrong (nesting without indentation).

ubsan replied:
Weeeeeeee

nikomatsakis replied:
@eddyb I don't think this ugly -- it's more "regular", if more "wordy" -- but yes the indentation is a bug. Probably worth filing an issue if we can isolate it.

ubsan replied:
This would be another case where a real rustfmt style guide would help. What is the correct thing to do here? ping @nrc

nrc replied:
https://github.com/rust-lang-nursery/rustfmt/issues/1132

SergioBenitez commented:
@nrc Would that be a problem though? That would only be in nightly, so it would be okay. Either someone is on stable and they don't get this change, or someone is only nightly/beta and they get this change and the macros.

nikomatsakis commented:
I've not read the code but I trust that between @eddyb, @nrc, and @cgswords this is looking good. However, I agree with @nrc that we need a feature gate (and some tests). Many attributes are not as thorough as they should be about checking for random extra attributes and arguments and so forth, I suspect.

nrc commented:
@SergioBenitez I'm not sure I understand what you're saying here - this feature will ride the trains, so in 12 weeks or so it will be on stable, let's say that I add an attribute today and forget to forbid use of lits, that gets stabilised quickly, and once it does (say in 24 weeks) it hits stable, then users on stable will be able to use this feature with that attribute.

jseyfried commented on src/librustc/hir/check_attr.rs:
```diff
@@ -53,34 +53,39 @@ impl<'a> CheckAttrVisitor<'a> {
+                            "attribute should be applied to struct"
+                        } else {
+                            continue
+                        }
```
nit: formatting (incorrect indent)

SergioBenitez replied:
This is also `rustfmt`. Doesn't look wrong to me though.

eddyb replied:
What it's doing is this:
```rust
                "packed" |
                    "simd" => {
                        if target != Target::Struct {
```
The arm body should not be indented twice from the arm itself.

SergioBenitez replied:
Ah, I see it now. Actually, this may have been my editor and not `rustfmt`.

Is it better to have ``"packed" | "simd"` on one line, or is it better the way it was before?

jseyfried replied:
The arm would look nicer (imo) and have less rightward drift as
```rust
                "packed" | "simd" => {
                    if target != Target::Struct {
                        "attribute should be applied to struct"
                    } else {
                        continue
                    }
                }
```
than
```rust
                "packed" |
                    "simd" => {
                        if target != Target::Struct {
                           "attribute should be applied to struct"
                        } else {
                            continue
                        }
                    }
```
I haven't seen the style with the extra indent anywhere in the compiler before, but I suppose `rustfmt` recently adopted it...

Regardless, orthogonal formatting changes make reviewing more difficult -- for future reference, perhaps do a separate commit for `rustfmt` changes.

nrc replied:
I'm not sure why this is happening - it makes sense if there are *a lot* of patterns and one line arm bodies, but not here. I think there is a combination of a bug (why we split at `|`) and a bad heuristic (double indent).

nrc replied:
I'm not sure why this is happening - it makes sense if there are *a lot* of patterns and one line arm bodies, but not here. I think there is a combination of a bug (why we split at `|`) and a bad heuristic (double indent).

https://github.com/rust-lang-nursery/rustfmt/issues/1136

SergioBenitez commented:
With the addition of the feature flag and a run-pass-fulldeps test, I believe I've addressed all of the comments. Is there anything else that needs to be done for this PR to land?

jseyfried commented on src/libsyntax/attr.rs:
```diff
@@ -357,6 +464,26 @@ pub fn last_meta_item_value_str_by_name(items: &[P<MetaItem>], name: &str)
+        },
+        span: span
+    })
+}
```
Both of these functions are just the identity (neither calls `sort_meta_items`), which does the real work.

n.b. `sort_meta_items` appears to be unused (both by `rustc` and by any plugins on Github) so it and these two functions you added can probably be removed.

SergioBenitez replied:
Good note. I'll remove them.

jseyfried commented on src/libsyntax/attr.rs:
```diff
@@ -367,15 +494,7 @@ pub fn sort_meta_items(items: Vec<P<MetaItem>>) -> Vec<P<MetaItem>> {
-            span: span
-        }
-    })).collect()
+    v.into_iter().map(|(_, m)| sort_meta_item(m)).collect()
```
This used to recursively sort `MetaItemKind::List`s but now does nothing.

jseyfried replied:
(now that they can contain literals, sorting `MetaItemKind::List`s might not even make sense)

SergioBenitez replied:
Same as above.

bors commented:
:umbrella: The latest upstream changes (presumably #35764) made this pull request unmergeable. Please resolve the merge conflicts.

SergioBenitez commented:
Rebased on master.

nrc commented:
r+ with MetaListItem renamed and with the commits squashed. Thanks for the PR, awesome work!

SergioBenitez commented:
@nrc: Done.

eddyb commented:
@Manishearth r=nrc

bors commented:
:umbrella: The latest upstream changes (presumably #35979) made this pull request unmergeable. Please resolve the merge conflicts.

SergioBenitez commented:
Rebased.

nikomatsakis commented:
@bors r=nrc

bors commented:
:pushpin: Commit 8250a26 has been approved by `nrc`

<!-- @bors r=nrc 8250a26b5bcea9190ac63e756c35d8a54bf9da0c -->

nikomatsakis commented:
r? @nrc

jseyfried commented:
This will land in the next breaking batch.
@bors r-

SergioBenitez commented:
@jseyfried Wasn't this supposed to be part of the previous batch? Don't see the harm in landing it now if so.

jseyfried commented:
@SergioBenitez Which previous batch? Afaik, none of the pending syntax-[breaking-change] PRs have landed yet (c.f. https://github.com/rust-lang/rust/issues/31645#issuecomment-239228452).

SergioBenitez commented:
@jseyfried: https://github.com/rust-lang/rust/pull/35979

jseyfried commented:
Those aren't syntax-[breaking-change]s though.
